cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/batch/ETLMapReduce.java

```diff
@@ -0,0 +1,201 @@
+  private static final Gson GSON = new Gson();
+
+  @Override
+  public void configure() {
+    setName("ETLMapReduce");
+    setDescription("MapReduce driver for Batch ETL Adapters");
+  }
+
+  @Override
+  public void beforeSubmit(MapReduceContext context) throws Exception {
+    Job job = context.getHadoopJob();
+
+    Preconditions.checkArgument(context.getRuntimeArguments().containsKey(Constants.ADAPTER_NAME));
+    Preconditions.checkArgument(context.getRuntimeArguments().containsKey(Constants.CONFIG_KEY));
+    Preconditions.checkArgument(context.getRuntimeArguments().containsKey(Constants.Source.SPECIFICATION));
+    Preconditions.checkArgument(context.getRuntimeArguments().containsKey(Constants.Sink.SPECIFICATION));
+    Preconditions.checkArgument(context.getRuntimeArguments().containsKey(Constants.Transform.SPECIFICATION));
+
+    ETLBatchConfig config = GSON.fromJson(context.getRuntimeArguments().get(Constants.CONFIG_KEY),
+                                          ETLBatchConfig.class);
+    StageSpecification sourceSpec = GSON.fromJson(context.getRuntimeArguments().get(Constants.Source.SPECIFICATION),
+                                                  StageSpecification.class);
+    StageSpecification sinkSpec = GSON.fromJson(context.getRuntimeArguments().get(Constants.Sink.SPECIFICATION),
+                                                StageSpecification.class);
+
+    prepareSource(sourceSpec, context, config.getSource());
+    prepareSink(sinkSpec, context, config.getSink());
+    prepareTransforms(context, config.getTransforms());
+
+    job.setMapperClass(ETLMapper.class);
+    job.setNumReduceTasks(0);
+  }
+
+  private void prepareSource(StageSpecification sourceSpec, MapReduceContext context, ETLStage sourceStage)
+    throws Exception {
+    BatchSource source = (BatchSource) Class.forName(sourceSpec.getClassName()).newInstance();
+    BatchSourceContext sourceContext = new MapReduceSourceContext(context, sourceStage, sourceSpec);
+    LOG.info("Source Stage : {}", sourceStage);
+    LOG.info("Source Class : {}", source.getClass().getName());
+    LOG.info("Specifications of Source : {}", sourceSpec);
+    source.prepareJob(sourceContext);
+  }
+
+  private void prepareSink(StageSpecification sinkSpec, MapReduceContext context, ETLStage sinkStage) throws Exception {
+    BatchSink sink = (BatchSink) Class.forName(sinkSpec.getClassName()).newInstance();
+    BatchSinkContext sinkContext = new MapReduceSinkContext(context, sinkStage, sinkSpec);
+    LOG.info("Sink Stage : {}", sinkStage);
+    LOG.info("Sink Class : {}", sink.getClass().getName());
+    LOG.info("Specifications of Sink : {}", sinkSpec);
+    sink.prepareJob(sinkContext);
+  }
+
+  private void prepareTransforms(MapReduceContext context, List<ETLStage> transformStages) {
+    Job job = context.getHadoopJob();
+
+    //Set list of Transform specifications to be used in the Mapper
+    job.getConfiguration().set(Constants.Transform.SPECIFICATION,
```

Comment:
ok spec is generated by the stage and config comes from the outside. Not clear if we really need spec, but keeping it for now and will remove at the end if it is really not used anywhere